Index: rootdev/rootdev.c
===================================================================
--- rootdev.orig/rootdev.c
+++ rootdev/rootdev.c
@@ -173,7 +173,8 @@ static int match_sysfs_device(char *name
       break;
 
     /* Recurse one level for devices that may have a matching partition. */
-    if (major(found_devt) == major(*dev) && minor(*dev) > minor(found_devt)) {
+    if ((major(found_devt) == major(*dev) && minor(*dev) > minor(found_devt)) ||
+         !strncmp(entry->d_name, "loop", 4)) {
       sprintf(working_path, "%s/%s", basedir, entry->d_name);
       found = match_sysfs_device(name, name_len, working_path, dev, depth + 1);
       if (found > 0)
